src/main/kotlin/org/rust/lang/core/types/infer/MemoryCategorization.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.types.infer
+
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.BindByValue
+import org.rust.lang.core.psi.ext.containerExpr
+import org.rust.lang.core.psi.ext.kind
+import org.rust.lang.core.psi.ext.mutability
+import org.rust.lang.core.types.inference
+import org.rust.lang.core.types.ty.Mutability
+import org.rust.lang.core.types.ty.TyPointer
+import org.rust.lang.core.types.ty.TyReference
+import org.rust.lang.core.types.type
+
+
+sealed class PointerKind(val mutability: Mutability)
+class BorrowedPointer(mutability: Mutability) : PointerKind(mutability)
+class UnsafePointer(mutability: Mutability) : PointerKind(mutability)
+
+
+enum class MutabilityCategory {
+    Immutable, Declared, Inherited;
+
+    companion object {
+        fun valueOf(mutability: Mutability): MutabilityCategory {
+            return when (mutability) {
+                Mutability.IMMUTABLE -> MutabilityCategory.Immutable
+                Mutability.MUTABLE -> MutabilityCategory.Declared
+            }
+        }
+    }
+
+    fun inherit(): MutabilityCategory {
+        return when (this) {
+            MutabilityCategory.Immutable -> MutabilityCategory.Immutable
+            MutabilityCategory.Declared, MutabilityCategory.Inherited -> MutabilityCategory.Inherited
+        }
+    }
+
+    val isMutable: Boolean get() {
+        return when (this) {
+            MutabilityCategory.Immutable -> false
+            MutabilityCategory.Declared, MutabilityCategory.Inherited -> true
+        }
+    }
+}
+
+val RsExpr.mutabilityCategory: MutabilityCategory? get() {
```

Comment:
I think that this piece of code may be rewritten like this: 
```kotlin
val RsExpr.mutabilityCategory: MutabilityCategory? get() {
    return when (this) {
        is RsUnaryExpr -> this.unaryExprMutability

        is RsDotExpr -> this.dotExpressionMutability

        is RsIndexExpr -> this.containerExpr?.mutabilityCategory?.inherit()

        is RsPathExpr -> this.pathMutability

        is RsParenExpr -> this.expr.mutabilityCategory

        else -> null
    }
}

private val RsUnaryExpr.unaryExprMutability: MutabilityCategory?
    get() {
        val expr = expr ?: return null
        if (mul == null) return null
        val pointer = expr.type.pointerKind ?: return null

        return MutabilityCategory.valueOf(pointer.mutability)
    }

private val Ty.pointerKind: PointerKind?
    get() = when (this) {
        is TyPointer -> UnsafePointer(mutability)
        is TyReference -> BorrowedPointer(mutability)
        else -> null
    }


private val RsDotExpr.dotExpressionMutability: MutabilityCategory?
    get() {
        val type = expr.type
        val isDeref = expr.inference?.adjustments?.get(this)?.any { it.kind == Adjust.DEREF } ?: false

        return if (isDeref && type is TyReference) {
            MutabilityCategory.valueOf(type.mutability)
        } else {
            expr.mutabilityCategory?.inherit()
        }
    }

private val RsPatBinding.patBindingMutability: MutabilityCategory
    get() {
        return if (kind is BindByValue && mutability.isMut) {
            MutabilityCategory.Declared
        } else {
            MutabilityCategory.Immutable
        }
    }

private val RsPathExpr.pathMutability: MutabilityCategory?
    get() {
        val declaration = path.reference.resolve() ?: return null

        // this brings false-negative, because such variable should has Immutable category:
        // let x; x = 1;
        // x = 2; <-- `x` is immutable, but it determined as mutable
        //
        // so it would be replaced with some kind of data-flow analysis
        if (declaration.hasNoAssignment()) return MutabilityCategory.Declared

        return when (declaration) {
            is RsConstant, is RsEnumVariant, is RsStructItem, is RsFunction -> MutabilityCategory.Declared

            is RsPatBinding -> declaration.patBindingMutability

            is RsSelfParameter -> MutabilityCategory.valueOf(declaration.mutability)

            else -> null
        }
    }

private fun RsElement.hasNoAssignment(): Boolean {
    val letExpr = ancestorStrict<RsLetDecl>()
    return letExpr != null && letExpr.eq == null
}
```

It is the same thing; basically I just eliminated different types of returns and moved logic to separate properties
It seems more readable that way IMO

Reply:
:+1: but I recommend to use a regular functions without receiver instead of properties and leave as a property just `val RsExpr.mutabilityCategory`

Reply:
@vlad20012 if you say so :sweat_smile: 
I am still not sure when to use propertie or functions when it comes to extensions in kotlin

Reply:
@fedochet, @vlad20012  Great remark, thank you! Also what do you think about using visitor pattern instead of just functions/properties?

Reply:
@ortem it depends from how many possible operations on this hierarchy will you actually need
I think that if you will need another one, you may refactor current approach and use visitors

Reply:
You mean this? :grinning: 
```kotlin
    private fun RsExpr.inferType(expected: Ty? = null): Ty {
        if (ctx.isTypeInferred(this)) error("Trying to infer expression type twice")

        val ty = when (this) {
            is RsPathExpr -> inferPathExprType(this)
            is RsStructLiteral -> inferStructLiteralType(this, expected)
            is RsTupleExpr -> inferRsTupleExprType(this, expected)
            is RsParenExpr -> this.expr.inferType(expected)
            is RsUnitExpr -> TyUnit
            is RsCastExpr -> inferCastExprType(this)
            is RsCallExpr -> inferCallExprType(this, expected)
            is RsDotExpr -> inferDotExprType(this, expected)
            is RsLitExpr -> inferLitExprType(this, expected)
            is RsBlockExpr -> this.block.inferType(expected)
            is RsIfExpr -> inferIfExprType(this, expected)
            is RsLoopExpr -> inferLoopExprType(this)
            is RsWhileExpr -> inferWhileExprType(this)
            is RsForExpr -> inferForExprType(this)
            is RsMatchExpr -> inferMatchExprType(this, expected)
            is RsUnaryExpr -> inferUnaryExprType(this, expected)
            is RsBinaryExpr -> inferBinaryExprType(this)
            is RsTryExpr -> inferTryExprType(this)
            is RsArrayExpr -> inferArrayType(this, expected)
            is RsRangeExpr -> inferRangeType(this)
            is RsIndexExpr -> inferIndexExprType(this)
            is RsMacroExpr -> inferMacroExprType(this)
            is RsLambdaExpr -> inferLambdaExprType(this, expected)
            is RsRetExpr -> inferRetExprType(this)
            is RsBreakExpr -> inferBreakExprType(this)
            is RsContExpr -> TyNever
            else -> TyUnknown
        }

        ctx.writeExprTy(this, ty)
        return ty
    }

```

Reply:
Yep